### PR TITLE
Vc/firmware set validation attribute

### PR DESCRIPTION
**Pull request by DoctorVin**
This is what I'm thinking about for a FirmwareValidation API endpoint. The only thing I think might be controversial is my choice of return for it (204/No Content vs 201/Created). I went with the former because while we need an id for internal database reasons, it doesn't make sense (to me) as the consumer of an API to use this id to do anything. If I wanted to look up whether a given firmware set was validated (and when and on which machine) I'm never going to search for the id of the validation record, I'd want to use firmware set id, or some fact about the server it was run on like vendor or model (or sku :trollface: )

Left undone here is the presentation of the firmware-validation. We don't have requirements for that yet, and lacking this doesn't block our ability to have something validate firmware. We can discuss presentation of that data in a different PR.

### CODE REVIEW

File: pkg/api/v1/router_firmware_set_test.go [hunk 1]
```diff
@@ -979,3 +980,38 @@ func TestIntegrationServerComponentFirmwareSetRemoveFirmware(t *testing.T) {
...
+	// Updating an existing validation is kosher.
+	err = s.Client.ValidateFirmwareSet(context.TODO(), srvID, setID, time.Now())
+	require.NoError(t, err, "update with fixture server id failed")
+
+	// Validate the same firmware with a non-fixture server id -- expect success
+	err = s.Client.ValidateFirmwareSet(context.TODO(), uuid.New(), setID, time.Now())
+	require.NoError(t, err, "update with unregistered server id failed")
+
+	// bogus firmware set id
+	err = s.Client.ValidateFirmwareSet(context.TODO(), uuid.New(), uuid.New(), time.Now())
```
Comment (joelrebel) [inline pkg/api/v1/router_firmware_set_test.go:1015]: why does this endpoint accept bogus server/firmware set IDs?

File: pkg/api/v1/router_firmware_set.go [hunk 2]
```diff
@@ -698,3 +701,89 @@ func (r *Router) firmwareSetDeleteMappingTx(ctx context.Context, _ *models.Compo
...
+				zap.Error(updErr),
+				zap.String("firmware.set", payload.FirmwareSet.String()),
+				zap.String("target.server", payload.TargetServer.String()),
+			).Warn("updating existing firmware validation record")
+			metrics.DBError("update firmware validation")
+			doRollback = true
+			dbErrorResponse2(c, "update firmware validation", updErr)
+			return
+		}
+	case sql.ErrNoRows:
```
Comment (joelrebel) [inline pkg/api/v1/router_firmware_set.go:758]: using `errors.Is()` would be ideal here, incase the error returned was wrapped

File: pkg/api/v1/router_firmware_set_test.go (hunk 1 above)
Comment (DoctorVin) [reply pkg/api/v1/router_firmware_set_test.go:1015]: The test is specifically that a bogus firmware set id triggers an error (foreign key violation) if there’s no corresponding firmware set. The server case is less tractable in testing, with some weird edge cases that arise from the consequence of clearing the servers table between test runs. The only arrangement that would work here is `ON DELETE CASCADE` which is not acceptable operationally. While I’m fine with cascading a delete because we removed the associated firmware set (so any facts about its validation are moot), we should not cascade if the server is removed, and restricting the delete breaks our test assumptions. 

In the end, server id is a placeholder for some other set of facts that we have yet to specify. It’s not a great solution, but it’s better than nothing.

File: pkg/api/v1/router_firmware_set.go (hunk 2 above)
Comment (jakeschuurmans) [reply pkg/api/v1/router_firmware_set.go:758]: SQLboiler doesnt wrap sql.ErrNoRows specifically
from
`	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, sql.ErrNoRows
		}
		return nil, errors.Wrap(err, "models: failed to execute a one query for bios_config_components")
	}`

File: internal/models/firmware_set_validation_facts.go [hunk 3]
```diff
@@ -0,0 +1,1097 @@
...
+
+// One returns a single firmwareSetValidationFact record from the query.
+func (q firmwareSetValidationFactQuery) One(ctx context.Context, exec boil.ContextExecutor) (*FirmwareSetValidationFact, error) {
+	o := &FirmwareSetValidationFact{}
+
+	queries.SetLimit(q.Query, 1)
+
+	err := q.Bind(ctx, exec, o)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
```
Comment (jakeschuurmans) [inline internal/models/firmware_set_validation_facts.go:322]: @joelrebel 

Here you can see how sqlboil handles ErrNoRows.

File: pkg/api/v1/router_firmware_set.go (hunk 2 above)
Comment (DoctorVin) [reply pkg/api/v1/router_firmware_set.go:758]: Adding `errors.Is()` here doesn't hurt and doesn't require a lot of effort.

**Review by jakeschuurmans — APPROVED** [review]
